### ISSUE TITLE
Prow tests do not GC resources: SA, clusters, etc.

### ISSUE THREAD

**Issue by Ark-kun** [kubeflow/pipelines#193]
See these Prow logs:
https://gubernator.k8s.io/pr/kubeflow/pipelines/155
https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/155/presubmit-sample-test/6/ 
https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/155/presubmit-sample-test/8/

The tests are terminated without triggering the cluster deletion.
```
sample-test-7qbb6   30m       workflows.argoproj.io/phase=Running\nSleep for 20 seconds...\n'
Workflow sample-test-7qbb6 is not finished.
NAME                AGE       LABELS
sample-test-7qbb6   30m       workflows.argoproj.io/phase=Running
Sleep for 20 seconds...
+ sleep 20
{"component":"entrypoint","level":"error","msg":"Entrypoint received interrupt: terminated","time":"2018-11-09T01:24:44Z"}
{"component":"entrypoint","level":"error","msg":"Process did not exit before 15s grace period","time":"2018-11-09T01:24:59Z"}
```
Successfull deletion looks like this:

```
[34mrun-xgboost-tests[0m:	+ echo '\''Copy the test results to GCS gs://ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/sample_test/'\''
[34mrun-xgboost-tests[0m:	+ gsutil cp junit_SampleXGBoostOutput.xml gs://ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/sample_test/junit_SampleXGBoostOutput.xml
[34mrun-xgboost-tests[0m:	Copy the test results to GCS gs://ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/sample_test/
[34mrun-xgboost-tests[0m:	Copying file://junit_SampleXGBoostOutput.xml [Content-Type=application/xml]...
[34mrun-xgboost-tests[0m:	/ [0 files][    0.0 B/  421.0 B]                                                
[34mrun-xgboost-tests[0m:	/ [1 files][  421.0 B/  421.0 B]                                                
[34mrun-xgboost-tests[0m:	Operation completed over 1 objects/421.0 B.                                      '
+ [[ NAME                AGE       LABELS
sample-test-vqfjn   40m       workflows.argoproj.io/completed=true,workflows.argoproj.io/phase=Succeeded = *\p\h\a\s\e\=\F\a\i\l\e\d* ]]
+ printf Name: sample-test-vqfjn Namespace: default ServiceAccount: default Status: Succeeded Created: Fri Nov 09 02:08:25 +0000 '(41' minutes 'ago)' Started: Fri Nov 09 02:08:32 +0000 '(41' minutes 'ago)' Finished: Fri Nov 09 02:48:42 +0000 '(1' minute 'ago)' Duration: 40 minutes 10 seconds Parameters: commit-sha: 7280c8b777fc92834e0ccc329f84a9194fc78d9c test-results-gcs-dir: gs://ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/sample_test cluster-type: create-gke bootstrapper-image: gcr.io/ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/bootstrapper api-image: gcr.io/ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/api frontend-image: gcr.io/ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/frontend scheduledworkflow-image: gcr.io/ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/scheduledworkflow persistenceagent-image: gcr.io/ml-pipeline-test/7280c8b777fc92834e0ccc329f84a9194fc78d9c/persistenceagent dataflow-tft-image-suffix: dataflow-tft dataflow-predict-image-suffix: dataflow-predict dataflow-tfma-image-suffix: dataflow-tfma dataflow-tfdv-image-suffix: dataflow-tfdv dataproc-create-cluster-image-suffix: dataproc-create-cluster dataproc-delete-cluster-image-suffix: dataproc-delete-cluster dataproc-analyze-image-suffix: dataproc-analyze dataproc-transform-image-suffix: dataproc-transform dataproc-train-image-suffix: dataproc-train dataproc-predict-image-suffix: dataproc-predict kubeflow-dnntrainer-image-suffix: kubeflow-dnntrainer kubeflow-deployer-image-suffix: kubeflow-deployer local-confusionmatrix-image-suffix: local-confusionmatrix local-roc-image-suffix: local-roc sample-tests-image-suffix: sample-tests '[39mSTEP[0m' PODNAME DURATION MESSAGE '[32m✔[0m' sample-test-vqfjn '├---[32m✔[0m' get-project sample-test-vqfjn-3395074898 1m '├-·-[32m✔[0m' build-dataflow-predict-image sample-test-vqfjn-169935994 11m '|' '├-[32m✔[0m' build-dataflow-tfdv-image sample-test-vqfjn-4042166549 16m '|' '├-[32m✔[0m' build-dataflow-tfma-image sample-test-vqfjn-4292335701 15m '|' '├-[32m✔[0m' build-dataflow-tft-image sample-test-vqfjn-1833216705 11m '|' '├-[32m✔[0m' build-dataproc-analyze-image sample-test-vqfjn-1339646615 10m '|' '├-[32m✔[0m' build-dataproc-create-cluster-image sample-test-vqfjn-1699863658 12m '|' '├-[32m✔[0m' build-dataproc-delete-cluster-image sample-test-vqfjn-3385359973 10m '|' '├-[32m✔[0m' build-dataproc-predict-image sample-test-vqfjn-1939206898 10m '|' '├-[32m✔[0m' build-dataproc-train-image sample-test-vqfjn-3913764689 12m '|' '├-[32m✔[0m' build-dataproc-transform-image sample-test-vqfjn-3758084383 9m '|' '├-[32m✔[0m' build-kubeflow-deployer-image sample-test-vqfjn-2533420374 11m '|' '├-[32m✔[0m' build-kubeflow-dnntrainer-image sample-test-vqfjn-613574227 13m '|' '├-[32m✔[0m' build-local-confusion-matrix sample-test-vqfjn-3526061026 16m '|' '├-[32m✔[0m' build-local-roc sample-test-vqfjn-230331776 16m '|' '└-[32m✔[0m' build-sample-tests-image sample-test-vqfjn-3561924792 9m '├---[32m✔[0m' deploy-ml-pipeline sample-test-vqfjn-765548532 4m '└-·-[32m✔[0m' run-condition-tests sample-test-vqfjn-38881280 5m '├-[32m✔[0m' run-exit-handler-tests sample-test-vqfjn-560939926 5m '├-[32m✔[0m' run-immediate-value-tests sample-test-vqfjn-2645725244 3m '├-[32m✔[0m' run-parallel-join-tests sample-test-vqfjn-203305695 3m '├-[32m✔[0m' run-sequential-tests sample-test-vqfjn-1799142840 3m '├-[32m✔[0m' run-tf-training-tests sample-test-vqfjn-2908448038 9m '├-[32m✔[0m' run-tfx-tests sample-test-vqfjn-1806710279 10m '└-[32m✔[0m' run-xgboost-tests sample-test-vqfjn-986370485 17m
Name:+ exit 0
+ delete_cluster
+ echo 'Delete cluster...'
Delete cluster...
+ gcloud container clusters delete sampletest-7280c8b777-18843 --async
```

**Comment by gaoning777** [comment]
added the signal capture in the test infra: https://github.com/kubeflow/pipelines/pull/863

**Comment by IronPan** [comment]
I see this still happening in some of the recent tests
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/1061/kubeflow-pipeline-e2e-test/1111434972578713600/build-log.txt

**Comment by IronPan** [comment]
This might happen when two commits are submitting closely. Prow terminate the first one without test having a chance to clean up the cluster.

**Comment by IronPan** [comment]
this seem to be the behavior of prow 
https://github.com/kubernetes/test-infra/blob/91186ea67814300ed42df1188bdd2357d1d86bb2/prow/entrypoint/run.go#L163

**Comment by IronPan** [comment]
This is causing the service account and role bindings to be accumulated in the test project, hitting the upper limit of the project and blocking subsequent tests.

**Comment by gaoning777** [comment]
Any suggestions how we can auto reclaim the resources when two commits are submitted around the same time?

**Comment by Ark-kun** [comment]
> Any suggestions how we can auto reclaim the resources when two commits are submitted around the same time?

Unless Prow sends some notification to the running process, there might be no way to proactively clean the resources. But we can run a GC script every day.

**Comment by Bobgy** [comment]
We are on maximum limit for service account again:
```
ERROR: (gcloud.deployment-manager.deployments.create) Error in Operation [operation-1565683469235-58ffb1703b425-a848a1f1-f8a82d93]: errors:
- code: RESOURCE_ERROR
  location: /deployments/e2e-87cf16c-16595/resources/e2e-87cf16c-16595-vm
  message: '{"ResourceType":"iam.v1.serviceAccount","ResourceErrorCode":"429","ResourceErrorMessage":{"code":429,"message":"Maximum
    number of service accounts on project reached.","status":"RESOURCE_EXHAUSTED","details":[{"@type":"type.googleapis.com/google.rpc.RetryInfo","retryDelay":"86401s"}],"statusMessage":"Too
    Many Requests","requestPath":"https://iam.googleapis.com/v1/projects/ml-pipeline-test/serviceAccounts","httpMethod":"POST"}}'
```

**Comment by Bobgy** [comment]
I made a request to increase quota from 100 to 200. I hope they can approve this so we are not blocked for now.

**Comment by numerology** [comment]
Yes it happens from time to time. Current temp fix is to manually delete those service account.

**Comment by Ark-kun** [comment]
I think there is bug in the Kubeflow cleanup code so that the SAs are not cleaner up at all.

**Comment by numerology** [comment]
Marked as fixed by #1857 
Feel free to re-open if similar problems pop up again
